### PR TITLE
fix: check all examples for 'margin' in DataCollatorForPreference

### LINKED ISSUE #5539: fix: DataCollatorForRewardModelingDataset checks margin only on examples[0], causing silent data loss after shuffle
## Problem

In `trl/trainer/reward_trainer.py`, `DataCollatorForRewardModelingDataset.torch_call()` checks for the `"margin"` key only on `examples[0]`:

```python
if "margin" in examples[0]:
    margins = torch.tensor([example["margin"] for example in examples], dtype=torch.float)
```

This has two failure modes:

1. **Silent data loss**: After dataset shuffling, if the first example lacks `"margin"` but subsequent examples have it, margins are silently dropped and the margin-aware reward loss is skipped with no warning.
2. **Runtime `KeyError`**: If the first example has `"margin"` but a later one does not (inconsistent dataset), the list comprehension crashes.

## Expected Behavior

All examples should be checked for the `"margin"` key. Missing margins should default to `0.0` gracefully.

## Fix

```python
if any("margin" in example for example in examples):
    margins = torch.tensor([example.get("margin", 0.0) for example in examples], dtype=torch.float)
```

### PR DESCRIPTION
**Pull request by antoinsader**
# What does this PR do?

Fixes #5539 

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case. https://github.com/huggingface/trl/issues/5539
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [x] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Low Risk**
> Small, localized change to data collation logic with added test coverage; primary risk is minor behavior change for mixed-margin batches.
> 
> **Overview**
> Fixes `DataCollatorForPreference` to detect `margin` across *all* examples in a batch (instead of only `examples[0]`) and to fill missing per-example margins with `0.0` when any margins are present.
> 
> Adds a regression test ensuring mixed batches (some examples with `margin`, some without) produce a `margin` tensor aligned with the batch and defaulted for missing entries.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 3069ceb8df88b367147f0deb3ac382354bd2620d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/reward_trainer.py [hunk 1]
```diff
@@ -201,8 +201,9 @@ def torch_call(self, examples: list[dict[str, Any]]) -> dict[str, Any]:
         # Convert to tensor
         chosen_ids = [torch.tensor(example["chosen_ids"]) for example in examples]
         rejected_ids = [torch.tensor(example["rejected_ids"]) for example in examples]
-        if "margin" in examples[0]:
-            margins = torch.tensor([example["margin"] for example in examples], dtype=torch.float)
+        if any("margin" in example for example in examples):
+            margins = torch.tensor([example["margin"] if "margin" in example else 0.0 for example in examples], dtype=torch.float)
```
[the remark this answers is not in the document]
Comment (antoinsader) [reply trl/trainer/reward_trainer.py:205]: I fixed it by creating a flag to check "has_margin" and use the flag for the condition checking.
